// @progress/kendo-react-editor/dist/npm/tools/insertLink.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinkToolNS = void 0;
var React = require("react");
var kendo_react_buttons_1 = require("@progress/kendo-react-buttons");
var kendo_editor_common_1 = require("@progress/kendo-editor-common");
var utils_1 = require("./utils");
var main_1 = require("../dialogs/main");
var kendo_react_intl_1 = require("@progress/kendo-react-intl");
var messages_1 = require("./../messages");
/**
 * @hidden
 */
var LinkToolNS;
(function (LinkToolNS) {
    /**
     * @hidden
     */
    LinkToolNS.createLinkTool = function (settings) {
        var Tool = /** @class */ (function (_super) {
            __extends(LinkTool, _super);
            function LinkTool() {
                var _this = _super !== null && _super.apply(this, arguments) || this;
                _this.state = {
                    openedDialog: false
                };
                _this.toggleDialog = function () {
                    _this.setState({
                        openedDialog: !_this.state.openedDialog
                    });
                };
                return _this;
            }
            LinkTool.prototype.render = function () {
                var _a = this.props, view = _a.view, render = _a.render, props = __rest(_a, ["view", "render"]);
                var state = view && view.state;
                var collapsed = state ? state.selection.empty : true;
                var linkMark = state && (0, kendo_editor_common_1.getMark)(state, state.schema.marks[settings.mark]);
                var localization = (0, kendo_react_intl_1.provideLocalizationService)(this);
                var insertHyperlink = settings.messages.insertHyperlink;
                var rendering = [(React.createElement(kendo_react_buttons_1.Button, __assign({ onClick: this.toggleDialog, key: "link", disabled: collapsed && !linkMark }, utils_1.onDownPreventDefault, { title: localization.toLanguageString(insertHyperlink, messages_1.messages[insertHyperlink]) }, settings.props, props))), (this.state.openedDialog && view && (React.createElement(main_1.EditorDialogs.InsertLinkDialog, { key: "insertLinkDialog", settings: settings, dir: props.dir, view: view, onClose: this.toggleDialog })) || null)];
                return render ? render.call(undefined, rendering, { view: view }) : rendering;
            };
            return LinkTool;
        }(React.Component));
        (0, kendo_react_intl_1.registerForLocalization)(Tool);
        return Tool;
    };
})(LinkToolNS = exports.LinkToolNS || (exports.LinkToolNS = {}));
