// @progress/kendo-react-editor/dist/npm/tools/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.userSelectNone = exports.itemRender = exports.formatString = exports.onDownPreventDefault = void 0;
var React = require("react");
/**
 * @hidden
 */
var onDownPreventDefault = {
    onMouseDown: function (e) { return e.preventDefault(); },
    onPointerDown: function (e) { return e.preventDefault(); }
};
exports.onDownPreventDefault = onDownPreventDefault;
/**
 * @hidden
 */
var userSelectNone = {
    userSelect: 'none'
};
exports.userSelectNone = userSelectNone;
/**
 * @hidden
 */
var formatString = function (input) {
    var args = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        args[_i - 1] = arguments[_i];
    }
    args.forEach(function (value, index) {
        input = input.replace('{' + index + '}', String(value));
    });
    return input;
};
exports.formatString = formatString;
/**
 * @hidden
 */
var itemRender = function (li, itemProps) {
    return itemProps.dataItem.style ? React.cloneElement(li, li.props, (React.createElement("span", { style: itemProps.dataItem.style }, li.props.children))) : li;
};
exports.itemRender = itemRender;
