// @progress/kendo-react-editor/dist/npm/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EditorDialogs = exports.EditorUtils = exports.EditorToolsSettings = exports.EditorTools = exports.Editor = exports.ProseMirror = void 0;
var Editor_1 = require("./Editor");
Object.defineProperty(exports, "Editor", { enumerable: true, get: function () { return Editor_1.Editor; } });
var main_1 = require("./tools/main");
Object.defineProperty(exports, "EditorTools", { enumerable: true, get: function () { return main_1.EditorTools; } });
var toolsSettings_1 = require("./config/toolsSettings");
Object.defineProperty(exports, "EditorToolsSettings", { enumerable: true, get: function () { return toolsSettings_1.EditorToolsSettings; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "EditorUtils", { enumerable: true, get: function () { return utils_1.EditorUtils; } });
var main_2 = require("./dialogs/main");
Object.defineProperty(exports, "EditorDialogs", { enumerable: true, get: function () { return main_2.EditorDialogs; } });
var kendo_editor_common_1 = require("@progress/kendo-editor-common");
/**
 * An object containing the content of ProseMirror packages used at the Editor component.
 *
 * "prosemirror-state",
 * "prosemirror-view",
 * "prosemirror-model",
 * "prosemirror-transform",
 * "prosemirror-commands",
 * "prosemirror-keymap",
 * "prosemirror-tables",
 * "prosemirror-schema-list",
 * "prosemirror-history",
 * "prosemirror-inputrules",
 * "prosemirror-dropcursor",
 * "prosemirror-gapcursor"
 */
exports.ProseMirror = {
    // prosemirror-state
    Selection: kendo_editor_common_1.Selection,
    SelectionRange: kendo_editor_common_1.SelectionRange,
    TextSelection: kendo_editor_common_1.TextSelection,
    NodeSelection: kendo_editor_common_1.NodeSelection,
    AllSelection: kendo_editor_common_1.AllSelection,
    EditorState: kendo_editor_common_1.EditorState,
    Plugin: kendo_editor_common_1.Plugin,
    PluginKey: kendo_editor_common_1.PluginKey,
    Transaction: kendo_editor_common_1.Transaction,
    // prosemirror-view
    Decoration: kendo_editor_common_1.Decoration,
    DecorationSet: kendo_editor_common_1.DecorationSet,
    EditorView: kendo_editor_common_1.EditorView,
    // prosemirror-model
    Node: kendo_editor_common_1.Node,
    ResolvedPos: kendo_editor_common_1.ResolvedPos,
    NodeRange: kendo_editor_common_1.NodeRange,
    Fragment: kendo_editor_common_1.Fragment,
    Slice: kendo_editor_common_1.Slice,
    ReplaceError: kendo_editor_common_1.ReplaceError,
    Mark: kendo_editor_common_1.Mark,
    Schema: kendo_editor_common_1.Schema,
    NodeType: kendo_editor_common_1.NodeType,
    MarkType: kendo_editor_common_1.MarkType,
    ContentMatch: kendo_editor_common_1.ContentMatch,
    DOMParser: kendo_editor_common_1.DOMParser,
    DOMSerializer: kendo_editor_common_1.DOMSerializer,
    // prosemirror-transform
    Transform: kendo_editor_common_1.Transform,
    Step: kendo_editor_common_1.Step,
    StepResult: kendo_editor_common_1.StepResult,
    joinPoint: kendo_editor_common_1.joinPoint,
    canJoin: kendo_editor_common_1.canJoin,
    canSplit: kendo_editor_common_1.canSplit,
    insertPoint: kendo_editor_common_1.insertPoint,
    dropPoint: kendo_editor_common_1.dropPoint,
    liftTarget: kendo_editor_common_1.liftTarget,
    findWrapping: kendo_editor_common_1.findWrapping,
    StepMap: kendo_editor_common_1.StepMap,
    MapResult: kendo_editor_common_1.MapResult,
    Mapping: kendo_editor_common_1.Mapping,
    AddMarkStep: kendo_editor_common_1.AddMarkStep,
    RemoveMarkStep: kendo_editor_common_1.RemoveMarkStep,
    ReplaceStep: kendo_editor_common_1.ReplaceStep,
    ReplaceAroundStep: kendo_editor_common_1.ReplaceAroundStep,
    replaceStep: kendo_editor_common_1.replaceStep,
    // prosemirror-commands
    deleteSelection: kendo_editor_common_1.deleteSelection,
    joinBackward: kendo_editor_common_1.joinBackward,
    selectNodeBackward: kendo_editor_common_1.selectNodeBackward,
    joinForward: kendo_editor_common_1.joinForward,
    selectNodeForward: kendo_editor_common_1.selectNodeForward,
    joinUp: kendo_editor_common_1.joinUp,
    joinDown: kendo_editor_common_1.joinDown,
    lift: kendo_editor_common_1.lift,
    newlineInCode: kendo_editor_common_1.newlineInCode,
    exitCode: kendo_editor_common_1.exitCode,
    createParagraphNear: kendo_editor_common_1.createParagraphNear,
    liftEmptyBlock: kendo_editor_common_1.liftEmptyBlock,
    splitBlock: kendo_editor_common_1.splitBlock,
    splitBlockKeepMarks: kendo_editor_common_1.splitBlockKeepMarks,
    selectParentNode: kendo_editor_common_1.selectParentNode,
    selectAll: kendo_editor_common_1.selectAll,
    wrapIn: kendo_editor_common_1.wrapIn,
    setBlockType: kendo_editor_common_1.setBlockType,
    toggleMark: kendo_editor_common_1.toggleMark,
    autoJoin: kendo_editor_common_1.autoJoin,
    chainCommands: kendo_editor_common_1.chainCommands,
    pcBaseKeymap: kendo_editor_common_1.pcBaseKeymap,
    macBaseKeymap: kendo_editor_common_1.macBaseKeymap,
    baseKeymap: kendo_editor_common_1.baseKeymap,
    // prosemirror-history
    history: kendo_editor_common_1.history,
    undo: kendo_editor_common_1.undo,
    redo: kendo_editor_common_1.redo,
    undoDepth: kendo_editor_common_1.undoDepth,
    redoDepth: kendo_editor_common_1.redoDepth,
    // prosemirror-inputrules
    InputRule: kendo_editor_common_1.InputRule,
    inputRules: kendo_editor_common_1.inputRules,
    undoInputRule: kendo_editor_common_1.undoInputRule,
    emDash: kendo_editor_common_1.emDash,
    ellipsis: kendo_editor_common_1.ellipsis,
    openDoubleQuote: kendo_editor_common_1.openDoubleQuote,
    closeDoubleQuote: kendo_editor_common_1.closeDoubleQuote,
    openSingleQuote: kendo_editor_common_1.openSingleQuote,
    closeSingleQuote: kendo_editor_common_1.closeSingleQuote,
    smartQuotes: kendo_editor_common_1.smartQuotes,
    wrappingInputRule: kendo_editor_common_1.wrappingInputRule,
    textblockTypeInputRule: kendo_editor_common_1.textblockTypeInputRule,
    // prosemirror-keymap
    keymap: kendo_editor_common_1.keymap,
    keydownHandler: kendo_editor_common_1.keydownHandler,
    // prosemirror-schema-list
    orderedList: kendo_editor_common_1.orderedList,
    bulletList: kendo_editor_common_1.bulletList,
    listItem: kendo_editor_common_1.listItem,
    addListNodes: kendo_editor_common_1.addListNodes,
    wrapInList: kendo_editor_common_1.wrapInList,
    splitListItem: kendo_editor_common_1.splitListItem,
    liftListItem: kendo_editor_common_1.liftListItem,
    sinkListItem: kendo_editor_common_1.sinkListItem,
    // prosemirror-dropcursor
    dropCursor: kendo_editor_common_1.dropCursor,
    // prosemirror-gapcursor
    gapCursor: kendo_editor_common_1.gapCursor,
    // prosemirror-tables
    tableEditing: kendo_editor_common_1.tableEditing,
    fixTables: kendo_editor_common_1.fixTables,
    fixTablesKey: kendo_editor_common_1.fixTablesKey,
    cellAround: kendo_editor_common_1.cellAround,
    isInTable: kendo_editor_common_1.isInTable,
    selectionCell: kendo_editor_common_1.selectionCell,
    moveCellForward: kendo_editor_common_1.moveCellForward,
    inSameTable: kendo_editor_common_1.inSameTable,
    findCell: kendo_editor_common_1.findCell,
    colCount: kendo_editor_common_1.colCount,
    nextCell: kendo_editor_common_1.nextCell,
    removeColSpan: kendo_editor_common_1.removeColSpan,
    addColSpan: kendo_editor_common_1.addColSpan,
    columnIsHeader: kendo_editor_common_1.columnIsHeader,
    tableNodes: kendo_editor_common_1.tableNodes,
    tableNodeTypes: kendo_editor_common_1.tableNodeTypes,
    CellSelection: kendo_editor_common_1.CellSelection,
    TableMap: kendo_editor_common_1.TableMap,
    tableEditingKey: kendo_editor_common_1.tableEditingKey,
    columnResizing: kendo_editor_common_1.columnResizing,
    columnResizingPluginKey: kendo_editor_common_1.columnResizingPluginKey,
    updateColumnsOnResize: kendo_editor_common_1.updateColumnsOnResize,
    selectedRect: kendo_editor_common_1.selectedRect,
    addColumn: kendo_editor_common_1.addColumn,
    addColumnBefore: kendo_editor_common_1.addColumnBefore,
    addColumnAfter: kendo_editor_common_1.addColumnAfter,
    deleteColumn: kendo_editor_common_1.deleteColumn,
    rowIsHeader: kendo_editor_common_1.rowIsHeader,
    addRow: kendo_editor_common_1.addRow,
    addRowBefore: kendo_editor_common_1.addRowBefore,
    addRowAfter: kendo_editor_common_1.addRowAfter,
    deleteRow: kendo_editor_common_1.deleteRow,
    mergeCells: kendo_editor_common_1.mergeCells,
    splitCell: kendo_editor_common_1.splitCell,
    splitCellWithType: kendo_editor_common_1.splitCellWithType,
    setCellAttr: kendo_editor_common_1.setCellAttr,
    toggleHeader: kendo_editor_common_1.toggleHeader,
    toggleHeaderRow: kendo_editor_common_1.toggleHeaderRow,
    toggleHeaderColumn: kendo_editor_common_1.toggleHeaderColumn,
    toggleHeaderCell: kendo_editor_common_1.toggleHeaderCell,
    goToNextCell: kendo_editor_common_1.goToNextCell,
    deleteTable: kendo_editor_common_1.deleteTable
};
