// @progress/kendo-editor-common/dist/es2015/table.js
import { addRowAfter as pmAddRowAfter, addRowBefore as pmAddRowBefore } from 'prosemirror-tables-ts';
import { rowTypeAttr } from './config/constants';
/**
 * Creates a table.
 * @returns Node
 */
export const createTable = (nodes, rows, columns) => {
    const { table, table_row, table_cell } = nodes;
    const tableRows = [];
    let cells;
    for (let r = 0; r < rows + 1; r++) {
        cells = [];
        for (let c = 0; c < columns + 1; c++) {
            cells.push(table_cell.createAndFill());
        }
        tableRows.push(table_row.createAndFill(undefined, cells));
    }
    return table.createAndFill(undefined, tableRows);
};
const closest = (selection, name) => {
    const pos = selection.$head;
    for (let i = pos.depth; i > 0; i--) {
        const node = pos.node(i);
        if (node.type.name === name) {
            return {
                pos: pos.before(i),
                node
            };
        }
    }
    return null;
};
export const addRowBefore = (state, dispatch) => {
    const cmdDispatch = dispatch && (tr => {
        const row = closest(tr.selection, 'table_row');
        const table = closest(tr.selection, 'table');
        if (row && table && row.node.attrs[rowTypeAttr]) {
            let index = 0;
            for (let i = 0; i < table.node.nodeSize; i++) {
                if (table.node.child(i).eq(row.node)) {
                    index = i;
                    break;
                }
            }
            const next = table.node.child(index - 1);
            const from = row.pos - next.nodeSize;
            tr.setNodeMarkup(from, undefined, { [rowTypeAttr]: row.node.attrs[rowTypeAttr] });
        }
        return dispatch(tr);
    });
    return pmAddRowBefore(state, cmdDispatch);
};
export const addRowAfter = (state, dispatch) => {
    const cmdDispatch = dispatch && (tr => {
        const row = closest(tr.selection, 'table_row');
        if (row && row.node.attrs[rowTypeAttr]) {
            const from = row.pos + row.node.nodeSize;
            tr.setNodeMarkup(from, undefined, { [rowTypeAttr]: row.node.attrs[rowTypeAttr] });
        }
        return dispatch(tr);
    });
    return pmAddRowAfter(state, cmdDispatch);
};
export { pmAddRowBefore, pmAddRowAfter };
